### PR TITLE
Stack safe default folds

### LINKED ISSUE #148: Make `foldrDefault`/`foldlDefault` stack-safe
Currently, `foldrDefault` and `foldlDefault` are not stack-safe, even when given a stack-safe `foldMap` implementation. @natefaubion proposes a stack-safe implementation strategy for these functions [here](https://discourse.purescript.org/t/stack-safe-default-foldr/2974/2).  Would you accept a pull request to make `foldrDefault` and `foldlDefault` stack-safe using this strategy?  This would almost certainly make the implementation slower than the current implementation, but remain _O(n)_.

### PR DESCRIPTION
**Pull request by ntwilson**
**Description of the change**

Implements stack-safe and performant versions of `foldrDefault` and `foldlDefault`.  Fixes #148.
There is code in here to benchmark `foldrDefault` and `foldlDefault`, but no way to compare with the old implementations.  You can do that comparison on my other branch [here](https://github.com/ntwilson/purescript-foldable-traversable/tree/benchmarks-for-stack-safe-folds). 

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [x] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Review by JordanMartinez — APPROVED** [review]
LGTM! Thanks!

**Comment by Masynchin** [comment]
PureScript of 0.15 is released. Seems that PR can be merged.

**Comment by JordanMartinez** [comment]
The merge conflicts need to be fixed.

**Review by JordanMartinez — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ Bugfixes:
 
 Other improvements:
 - Narrow down unnecessarily imprecise type of `mapWithIndexArray` (#145)
+- Make `foldrDefault` and `foldlDefault` stack safe (#148)
```
Comment (thomashoneyman) [inline CHANGELOG.md:27]: This should be moved to the UNRELEASED section

**Review by thomashoneyman — APPROVED** [review]

**Pull request merged**